### PR TITLE
optimized textwidth(::Char) for ASCII

### PR DESCRIPTION
**Pull request by stevengj**
For the common case of ASCII characters, I found in https://github.com/JuliaLang/julia/pull/55351/files#r1705746669 that `textwidth` can be sped up by ~~60%~~ 2.7x or more by adding a fast-path ASCII ~~lookup table~~ hard-coded rule.

The slowdown for non-ASCII characters seems negligible (1% or less).  Here, `textwidth2` is the new function, and `textwidth` is the old function (Julia 1.10.4, Apple M1 Pro):
```jl
julia> @btime textwidth('x'); @btime textwidth2('x')
  3.167 ns (0 allocations: 0 bytes)
  1.166 ns (0 allocations: 0 bytes)
1

julia> @btime textwidth('α'); @btime textwidth2('α')
  3.208 ns (0 allocations: 0 bytes)
  3.166 ns (0 allocations: 0 bytes)
1
```

### CODE REVIEW

**Comment by jakobnissen** [comment]
Can it be faster if the const lookup table is a Tuple? The reason is that const-ness at the moment only refers to the binding, so a const vector will do a triple load (load the underlying memoryref, then load the pointer, then load the offset) whereas an immutable Tuple might generate better code.

**Comment by stevengj** [comment]
@jakobnissen, good catch, a tuple table indeed seems to be faster:
```jl
julia> @btime textwidth('x'); @btime textwidth2('x'); @btime textwidth3('x');
  3.166 ns (0 allocations: 0 bytes)
  1.958 ns (0 allocations: 0 bytes)
  1.125 ns (0 allocations: 0 bytes)
```
(`textwidth` is the old version, `textwidth2` is the array version, and `textwidth3` is the tuple version)

**Comment by stevengj** [comment]
Another possibility would be to omit the table entirely and replace it with a couple of `if` statements, since the table is quite simple right now:
```jl
(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0)
```

**Comment by stevengj** [comment]
Looks like an unrelated failure in the REPL tests?

File: base/strings/unicode.jl [hunk 1]
```diff
@@ -256,6 +256,15 @@ julia> textwidth('⛵')
 ```
 """
 function textwidth(c::AbstractChar)
+    ismalformed(c) && return 1
+    i = codepoint(c)
+    i < 0x7f && return Int(i >= 0x20); # ASCII fast path
```
Comment (JeffBezanson) [inline base/strings/unicode.jl:261]:
```suggestion
    i < 0x7f && return Int(i >= 0x20) # ASCII fast path
```

File: base/strings/unicode.jl [hunk 2]
```diff
@@ -256,6 +256,15 @@ julia> textwidth('⛵')
...
+    Int(ccall(:utf8proc_charwidth, Cint, (UInt32,), i))
+end
+
+function textwidth(c::Char)
+    b = bswap(reinterpret(UInt32, c)) # from isascii(c)
+    b < 0x7f && return Int(b >= 0x20); # ASCII fast path
```
Comment (JeffBezanson) [inline base/strings/unicode.jl:267]:
```suggestion
    b < 0x7f && return Int(b >= 0x20) # ASCII fast path
```

**Review by JeffBezanson — APPROVED** [review]

**Pull request merged**